oak_functions/loader/src/server.rs

```diff
@@ -149,25 +148,54 @@ impl FixedSizeBodyPadder for Response {
     }
 }
 
+/// Trait for implementing extensions, to implement new native functionality.
+pub trait OakApiNativeExtension {
+    /// Similar to `invoke_index` in [`wasmi::Externals`], but may return a result to be
+    /// written into the memory of the `WasmState`.
+    fn invoke_index(
+        &self,
+        wasm_state: &WasmState,
```

Comment:
Is it possible to pass `&mut WasmState` here, so that the function can just write directly to memory? It seems odd that it can read directly from it, but it needs to return data via a custom mechanism in order for it to be written.

Reply:
I think it would have been possible if extensions were clonable, but I am not sure how to do that without making extensions clonable.

`invoke` is called inside `WasmState::invoke_index`, which borrows `self` mutably. Then somewhere at the end of `invoke_index`, an extension is retrieved from the hashmap: `self.extensions_indices.get(&index)`. When calling `invoke` on the retrieved extension, `self` is still borrowed, so it cannot be borrowed mutably again. 

I think it makes sense, it should not be possible to mutate `self` (by writing into its memory), while it is being borrowed (for accessing one of its extensions).

Reply:
The Extension is wrapped in `Arc`, so it is actually clonable. Fixed this in #2285.